### PR TITLE
V2.0.1

### PR DESCRIPTION
**Pull request by dlip**
- Rewrite templating to be more efficient
- Bugfix: Throw error when template variable not set

### CODE REVIEW

File: tests/templating.env [hunk 1]
```diff
@@ -1,2 +1,13 @@
...
+NAME=envy
+DOUBLE={{NAME}}-{{VERSION}}-pro
+DOUBLE_NO_CLOSE={{NAME}}-{{VERSION-pro
+DOUBLE_OPEN={{NAME-{{VERSION}}-pro
+NOT_TEMPLATE=envy-{{}}-pro
+NO_CLOSE_START={{-envy
+NO_CLOSE_MIDDLE=envy-{{-pro
+NO_CLOSE_END=envy-{{
+VAR_START={{VERSION}}-pro
+VAR_MIDDLE=envy-{{VERSION}}-pro
```
Comment (leontee) [inline tests/templating.env:11]: Will something like the below pass and substitute version as expected?

VAR_START={{{VERSION}}-pro}
VAR_MIDDLE=envy-{{{VERSION}}-pro}
VAR_END=envy-{{{VERSION}}}

File: envy.sh [hunk 2]
```diff
@@ -6,58 +6,64 @@ set -euo pipefail
...
-
-            line="${line//${LEFT_DELIM}$key${RIGHT_DELIM}/$val}"
-          done
+    # Read input char by char
+    while read -n1 CHAR; do
+        # If found {{
+        if [[ "${CHAR}" == "{" && "${LAST_CHAR}" == "{" ]]; then
+            # Remove extra {
+            RESULT="${RESULT::-1}"
+            local VAR_RESULT=""
```
Comment (leontee) [inline envy.sh:19]: Any reason why we choose prefix VAR_ here for local vars on the inner loop? It's ok, but was a bit confusing first read through

Comment (dlip) [reply envy.sh:19]: Yea it is a bit hard to follow, its an inner loop which is searching for the var name inside the brackets

File: tests/templating.env (hunk 1 above)
Comment (dlip) [reply tests/templating.env:11]: This all result in an error of `Error while parsing tempate, {VERSION not set` which is correct.
If you replace it with the correct version tag but keep the final } it is also ok
```
VAR_START={{VERSION}}-pro}
VAR_MIDDLE=envy-{{VERSION}}-pro}
VAR_END=envy-{{VERSION}}}
```
->
```
export VAR_END=envy-1.0.0}
export VAR_MIDDLE=envy-1.0.0-pro}
export VAR_START=1.0.0-pro}
```

Comment (leontee) [reply tests/templating.env:11]: I guess we read code differently. I'd have expected the below by default

```
VAR_START={{{VERSION}}-pro}
VAR_MIDDLE=envy-{{{VERSION}}-pro}
VAR_END=envy-{{{VERSION}}}
```
->
```
export VAR_END={envy-1.0.0}
export VAR_MIDDLE={envy-1.0.0-pro}
export VAR_START={1.0.0-pro}
```
I'm not too fussed, edge case :)

**Review by leontee — APPROVED** [review]

**Review by leontee — APPROVED** [review]
